app/bundles/PageBundle/Tests/EventListener/BuildJsSubscriberTest.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Mautic\PageBundle\Tests\EventListener;
+
+use Monolog\Logger;
+use Mautic\CoreBundle\Event\BuildJsEvent;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Mautic\PageBundle\EventListener\BuildJsSubscriber;
+use Mautic\CoreBundle\Twig\Helper\AssetsHelper;
+use Symfony\Component\Asset\Packages;
+use Mautic\CoreBundle\Helper\CoreParametersHelper;
+use Symfony\Component\Routing\RouterInterface;
+use Mautic\PageBundle\Helper\TrackingHelper;
+use Mautic\CoreBundle\Helper\PathsHelper;
+use Mautic\CoreBundle\CoreEvents;
+use PHPUnit\Framework\TestCase;
+
+class BuildJsSubscriberTest extends TestCase
+{
+    public function testBuildMauticJsReturnsGtagAnalytics()
+    {
+        $dispatcher = new EventDispatcher();
+        $subscriber = $this->getBuildJsSubscriber();
+
+        $dispatcher->addSubscriber($subscriber);
+
+        $event = new BuildJsEvent($this->getJsHeader(), false);
+
+        $dispatcher->dispatch($event, CoreEvents::BUILD_MAUTIC_JS);
+
+        $result = strpos($event->getJs(), 'https://www.googletagmanager.com/gtag/js?id=G-F3825DS9CD');
+        $this->assertNotEquals($result, false);
+    }
+
+    /**
+     * Get buildjs subscriber with mocked dependencies.
+     */
+    protected function getBuildJsSubscriber(): BuildJsSubscriber
+    {
+        /** @var Packages&MockObject $packagesMock */
+        $packagesMock = $this->createMock(Packages::class);
+
+        /** @var CoreParametersHelper&MockObject $coreParametersHelper */
+        $coreParametersHelper = $this->createMock(CoreParametersHelper::class);
+
+        $assetsHelperMock   = new AssetsHelper($packagesMock, $coreParametersHelper);
+        $pathsHelperMock  = $this->createMock(PathsHelper::class);
+        $pathsHelperMock->method('getSystemPath')->willReturn('http://localhost');
+        $assetsHelperMock->setPathsHelper($pathsHelperMock);
+
+        $trackingHelperMock   = $this->createMock(TrackingHelper::class);
+        $trackingHelperMock->method('displayInitCode')->willReturnCallback(function ($key) {
+            switch ($key) {
+                case 'google_analytics':
+                    return 'G-F3825DS9CD';
+                    break;
+            };
+            return false;
+        });
+
+        $routerHelperMock   = $this->createMock(RouterInterface::class);
+        $routerHelperMock->method('generate')->willReturn('http://localhost');
+
+        return new BuildJsSubscriber(
+            $assetsHelperMock,
+            $trackingHelperMock,
+            $routerHelperMock
+        );
+    }
+
+     /**
+     * Build a JS header for the Mautic embedded JS.
+     *
+     * @return string
+     */
+    protected function getJsHeader()
```

Comment:
```suggestion
    protected function getJsHeader(): string
```
How is such comment related to all this?

Reply:
What issues? There was no stan or cs issues when i made the pr last month.

Reply:
Also worth noting I didn't change the code you are talking about. I think your issues are unrelated.

Reply:
You are adding this code, so I don't get this:

> Also worth noting I didn't change the code you are talking about

I don't understand what is this method for. It returns some comment and I don't understand why.

Reply:
The PR updates analytics.js to gtag which is injected from the BuildJsEvent event which is called in the below file. 

https://github.com/mautic/mautic/blob/88baa803e4613531eac14473b218b0676e373314/app/bundles/CoreBundle/Controller/JsController.php#L24

It has the getJsHeader function in there I copy pasted the code from that for the test. I don't know why its like that Mautic just be like that.

Reply:
I see. But that's not important in your code, right? You can simply create the event object like
```php
$event = new BuildJsEvent('', false);
```
And it will have no effect on this test. But this could be all tested better with a functional test. Let me write it and push it so you can see how you can write a way smaller test that would actually test the feature.

Reply:
No idea what parameters BuildJsEvent needs. I just tested it works the way its used but if you have a better test then great.

Reply:
I don't want to push this as it's failing but this is the same as yours test just functional:
```php
<?php

declare(strict_types=1);

namespace Mautic\CoreBundle\Tests\Functional\Controller;

use Mautic\CoreBundle\Test\MauticMysqlTestCase;
use PHPUnit\Framework\Assert;
use Symfony\Component\HttpFoundation\Response;

final class JsControllerTest extends MauticMysqlTestCase
{
    protected function setUp(): void
    {
        $this->configParams['google_analytics'] = 'G-F3825DS9CD';
        parent::setUp();
    }
    
    public function testIndexActionRendersSuccessfully(): void
    {
        $this->client->request('GET', '/mtc.js');
        Assert::assertSame(Response::HTTP_OK, $this->client->getResponse()->getStatusCode());
        Assert::assertStringContainsString('https://www.googletagmanager.com/gtag/js?id=G-F3825DS9CD', $this->client->getResponse()->getContent());
    }
}
```
Does it make sense? You don't have to configure all the mocks and services. Just use the existing ones. Test them all and how they work together. Test what output you get for what input. That's a test you need when you refactor some code too. And you know that it will continue working and you don't have to refactor the unit tests together with the code. It adds much more value.

Now the question is why it's not adding the GA code. Do you know?

Reply:
Yea your test looks better. It adds the GA code only if the google_analytics config param is set which works for me in staging and worked on my test. I would assume that your test isn't setting the config param correctly.

Reply:
The actual param name is `google_analytics_id`. Not just `google_analytics_id`. And `google_analytics_trackingpage_enabled` must be enabled too. I pushed a commit to replace your test with the functional one.

Reply:
cool